Internal: Ubuntu support

maelk opened the pull request:
Do not merge! This is for internal review

This pull request adds support for Ubuntu environment

maelk commented on 02_configure_host.sh:
```diff
@@ -48,52 +48,59 @@ EOF
+OS=$(uname -a)
+if [[ $OS == *Ubuntu* ]]; then
+  # source ubuntu_bridge_network_configuration.sh
+  source ubuntu_bridge_network_configuration.sh
```
The Centos part could then also be split in its own script

namnx228 replied:
I also want to discuss with you guys about this. I tried to make as less change as possible, so it will be more easier for maintainer to review our code. For that reason, I tried to avoid changing to what they have done, and just adding new code.
What do you think about this ? Should we also modify what they have done to make the code easier to read ? Or maybe we can do it in the another pull request, so keep this pull request as familiar with the maintainer as possible ?

jan-est replied:
I would say that we should not do too much changes. We need to check how our changes are accepted in the community. Let's hope pull request generates discussion in the community. I think we are not offering "ultimate" solution, but contribution dev-env for further development?

maelk commented on Makefile:
```diff
@@ -1,7 +1,7 @@
 
 install_requirements:
-	./01_install_requirements.sh
+	./ubuntu_requirement_installation.sh
```
This will break it for CentOS, since you call apt for example in the ubuntu_* script

namnx228 replied:
Yes, I'm working on this case. I plan to put some bash code here to identify which OS is running

namnx228 replied:
I have made the dev-env running on both Ubuntu and CentOS at the moment. Now I will try to run the dev-env on a fresh new CentOS VM

namnx228 replied:
Please check the latest commits.

maelk commented on 02_configure_host.sh:
```diff
@@ -128,11 +135,7 @@ if [ "$MANAGE_BR_BRIDGE" == "y" ] ; then
   if [ "$ADDN_DNS" ] ; then
     echo "server=$ADDN_DNS" | sudo tee /etc/NetworkManager/dnsmasq.d/upstream.conf
```
Only for CentOS

namnx228 replied:
Yeah. I will update it.
Btw, do we need to configure dnsmasq, since I see that they have a dnsmasq running on a container ?

maelk replied:
This was not solved yet

maelk commented on 02_configure_host.sh:
```diff
@@ -128,11 +135,7 @@ if [ "$MANAGE_BR_BRIDGE" == "y" ] ; then
-  else
-    sudo systemctl restart NetworkManager
-  fi
+  sudo netplan apply
```
```suggestion
  OS=$(uname -a)
  if [[ $OS == *Ubuntu* ]]; then
    sudo netplan apply
  else
    if systemctl is-active --quiet NetworkManager; then
      sudo systemctl reload NetworkManager
    else
      sudo systemctl restart NetworkManager
    fi
  fi
```

maelk commented on ubuntu_requirement_installation.sh:
```diff
@@ -0,0 +1,126 @@
+
+##No bind-utils. It is for host, nslookop,..., no need in ubuntu
+sudo apt -y install \
+  ansible \
```
The ansible installed this way has issues. We should probably go with virtual environments

maelk commented on ubuntu_requirement_installation.sh:
```diff
@@ -0,0 +1,126 @@
+sudo pip install \
+  lolcat \
+  yq \
+  virtualbmc \
```
we could use the containerized version we worked on

kanwar-saad replied:
you need to install docker then

maelk replied:
or podman. There are anyways other containers that are running this way, like ironic, mariadb, httpd etc.

maelk commented on vm-setup/roles/virtbmc/tasks/setup_tasks.yml:
```diff
@@ -52,13 +52,22 @@
   become: true
 
+- name: Start the Virtual BMCs (virtualbmc >= 1.4.0+) on ubuntu
+  shell: vbmcd || true
```
A service based approach would be better, or using a containerized vbmc

namnx228 replied:
Virtual BMCs service is not available on Ubuntu, so containerized version should be better in this case. We can propose this approach in another issue in the future. At the moment, I believe that we should show the community a runnable version of dev-env, and get their feedback as well.

maelk replied:
ok, up to you

maelk commented on vm-setup/roles/virtbmc/tasks/teardown_tasks.yml:
```diff
@@ -10,9 +10,17 @@
+- name: Stop/disable the Virtual BMCs (virtualbmc >= 1.4.0+) on Ubuntu
+  when:
+    - ansible_distribution == 'Ubuntu' 
+  shell: pkill vbmcd || true
```
same here

maelk commented on ubuntu_requirement_installation.sh:
```diff
@@ -0,0 +1,126 @@
+  curl \
+  dnsmasq \
+  figlet \
+  golang \
```
this might be too old. Use the specific repo to install 1.12

namnx228 replied:
The only way to install the latest version of Golang (as far as my investigation) is to download the binary and move it to /usr/local/bin/. I will do it, but it is not a good solution in the long run.

maelk commented on ubuntu_requirement_installation.sh:
```diff
@@ -0,0 +1,126 @@
+  nmap \
+  patch \
+  psmisc \
+  python-pip \
```
All the python-* should be installed in a virtual env like ansible

namnx228 replied:
I'm also considering that should we use apt or pip to install Python packages. I consulted the ideas in this page: https://unix.stackexchange.com/questions/450008/pip-vs-package-manager-for-handling-python-packages. 

> The biggest disadvantage I see with using pip to install Python modules on your system, either as system modules or as user modules, is that your distribution’s package management system won’t know about them. This means that they won’t be used for any other package which needs them, and which you may want to install in the future (or which might start using one of those modules following an upgrade); you’ll then end up with both pip- and distribution-managed versions of the modules, which can cause issues (I ran into yet another instance of this recently). So your question ends up being an all-or-nothing proposition: if you only use pip for Python modules, you can no longer use your distribution’s package manager for anything which wants to use a Python module...

I think it makes sense.
How do you think about this problem ?

maelk replied:
If we use a virtual environment, we won't be impacted by or impacting the python distribution installed, we'll be completely independent, so we shouldn't hit such problems

namnx228 replied:
Ok, that makes sense. So why should we install all python-* in a virtual env ?

maelk reviewed (commented):
Please verify that you did not break anything by running it on CentOS

kanwar-saad commented on ubuntu_bridge_network_configuration.sh:
```diff
@@ -0,0 +1,88 @@
+     # the IP address here
+     IFCFG_PROVISIONING_FILE=/etc/netplan/provision-bridge.yaml
+     IFCFG_PROVISIONING=$(cat << EOF
+network:
```
I suggest to create a template file and fill the template once all ifs and buts are resolved.

namnx228 replied:
Do you get the latest version ? Those lines should not be in `ubuntu_bridge_network_configuration.sh`

kanwar-saad commented on ubuntu_bridge_network_configuration.sh:
```diff
@@ -0,0 +1,88 @@
+             sudo sed -i -e "s/\(.*\)dhcp4: no/\1dhcp4: yes/" $IFCFG_BAREMETAL_FILE
+             sudo systemctl restart network
+         else
+            sudo systemctl restart network
```
this can be taken out of if condition

kanwar-saad reviewed (commented):
managing OS specific configurations can be done in a more neater way but this approach is ok for now. we have to design it properly later if we way to spend some effort on it.

kanwar-saad commented on ubuntu_bridge_network_configuration.sh:
```diff
@@ -6,75 +6,26 @@ if [ "$MANAGE_PRO_BRIDGE" == "y" ]; then
-EOF
-     )
+     sudo brctl addbr provisioning
+     sudo ifconfig provisioning 172.22.0.1 netmask 255.255.255.0 up 
```
use ip command. ifconfig commands are deprecated now.

namnx228 replied:
Yes, ifconfig commands are deprecated on Ubuntu now, so I added `net-tools` package.
```sh
51   echo "Install net-tools package"
52   sudo apt -y install net-tools

```

maelk replied:
this should be changed for ip, even if we don't change it in the CentOS part, we should add proper code that does not immediately require changes

jan-est replied:
I have worked on this one! Got it working, but I think I do another pull request when this on is pulled to upstream?

kanwar-saad reviewed (commented):
its better to squash these commits into one... that is easier to review.

namnx228 commented:
> its better to squash these commits into one... that is easier to review.

How to do that ?

Xenwar commented on 02_configure_host.sh:
```diff
@@ -29,9 +29,17 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
+    originalGRP=$(id -g -n)
+    # A trick to add 'libvirt' group to user without reloginrelogin 
+    newgrp libvirt
+    newgrp $originalGRP
```
Some actions need to take place between the two newgrp commands. 
Or remove the second one as it only does restoring.

maelk commented on 02_configure_host.sh:
```diff
@@ -34,6 +34,12 @@ if ! id $USER | grep -q libvirt; then
   sudo systemctl restart libvirtd
 fi
 
+# Current user is not effectively in group `libvirt` yet, so give full permission to access `libvirt-sock`. The same permission is given on Centos by default.
```
line break

maelk commented on 02_configure_host.sh:
```diff
@@ -34,6 +34,12 @@ if ! id $USER | grep -q libvirt; then
+# Current user is not effectively in group `libvirt` yet, so give full permission to access `libvirt-sock`. The same permission is given on Centos by default.
+OS=$(uname -a)
+if [[ $OS == *Ubuntu* ]]; then
+  sudo chmod o+rwx /var/run/libvirt/libvirt-sock
```
I think this is a really bad idea! Can we work around the current user not being yet in libvirt with another way ? such as newgrp ?

namnx228 replied:
I have tried some workaround like using `newgrp`, but it doesn't work. `newgrp` spawns a subshell, and the script cannot continue to run from there, so the rest of scripts (lines of code below the command `newgrp`) cannot be executed. The reason is that the rest of is in the old shell, and we get stuck at the new shell.
Besides, on CentOS, `libvirt-sock` is given full permission, so we don't see this problem. 
Therefore, I think manually changing permission of this file is proper solution at the moment. After this pull request, we can create an issue about this bug, and discuss about a better solution.

maelk commented on disable_apparmor_driver_libvirtd.sh:
```diff
@@ -0,0 +1,8 @@
+foo="#security_driver = \"selinux\""
```
naming could be improved :p

maelk commented on disable_apparmor_driver_libvirtd.sh:
```diff
@@ -0,0 +1,8 @@
+sudo sed -i "s/$foo/$bar/g" /etc/libvirt/qemu.conf
+sudo sed -i "s/$foo2/$bar/g" /etc/libvirt/qemu.conf
+sudo systemctl restart libvirtd 
+sudo chmod o+rwx /var/run/libvirt/libvirt-sock
```
Again, I would not do that!

maelk commented on lib/common.sh:
```diff
@@ -67,14 +67,15 @@ if ! sudo -n uptime &> /dev/null ; then
-if [[ $(awk -F= '/^VERSION_ID=/ { print $2 }' /etc/os-release | tr -d '"' | cut -f1 -d'.') -ne 7 ]]; then
-  echo "Required CentOS 7 or RHEL 7"
+os_version=$(awk -F= '/^VERSION_ID=/ { print $2 }' /etc/os-release | tr -d '"' | cut -f1 -d'.')
+if [[ ${os_version} -ne 7 ]] && [[ ${os_version} -ne 18 ]]; then
```
maybe we should make the conditions tighter and like it to the OS

namnx228 replied:
Updated

maelk commented on ubuntu_install_requirements.sh:
```diff
@@ -0,0 +1,151 @@
+
+# Install pyenv
+if [[ $PATH != *pyenv* ]]; then
+  export PYENV_ROOT="$HOME/.pyenv"
```
we should rather make it relative to the current path IMO, to not mess up with the user's home directory

maelk commented on ubuntu_install_requirements.sh:
```diff
@@ -0,0 +1,151 @@
+  fi
+fi
+
+if [[  $(cat ~/.bashrc) != *PYENV_ROOT* ]]; then
```
do this whole part first

maelk commented on ubuntu_install_requirements.sh:
```diff
@@ -0,0 +1,151 @@
+if [[ $PATH != *pyenv* ]]; then
+  export PYENV_ROOT="$HOME/.pyenv"
+  export PATH="$PYENV_ROOT/bin:$PATH"
+  if command -v pyenv 1>/dev/null 2>&1; then
```
if you check for installation and install first, you don't need this check

namnx228 replied:
We still need it. In the case that users run `make` and get an error, and they run `make` again on the same shell. At the point, we have added `export PYENV_ROOT ...` in .bashrc file, but $PATH is still not updated. So, there is an error in that case.

maelk commented on ubuntu_install_requirements.sh:
```diff
@@ -0,0 +1,151 @@
+  if ! [ -d "$HOME/.pyenv" ] ; then
+     git clone git://github.com/yyuu/pyenv.git ~/.pyenv
+  fi
+  echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.bashrc
```
This cannot run twice or it will duplicate entries in the bashrc. I would rather not add anything in bashrc but make a small file to source when you want to work with pyenv

namnx228 replied:
This will not run twice as there is a check if `PYENV_ROOT` is inside .bashrc. At the moment, let keep it simple as adding something to bashrc file. We can improve it in the future.

maelk commented on ubuntu_install_requirements.sh:
```diff
@@ -0,0 +1,151 @@
+
+if [[  $(cat ~/.bashrc) != *PYENV_ROOT* ]]; then
+  if ! [ -d "$HOME/.pyenv" ] ; then
+     git clone git://github.com/yyuu/pyenv.git ~/.pyenv
```
the location should be relative to the current working directory IMO since we are automatically installing pyenv

namnx228 replied:
The same here. And I think it can be good if we can make pyenv available for other applications, so users don't need to reinstall it in the future when needing.

namnx228 committed:
Make dev-env support Ubuntu

Disable Apparmor driver of Libvirt so VMs can start
Update Makefile with OS detectable feature.
Use Ansible 2.8.2 in place of 2.5.2
Install Go 1.12 using apt 
Update previous commit
Install Pyenv and use Pip to install python packages
Fix error in pyenv installation
Fix error: current user doesn't have permission to access libvirt
change permission each time libvirtd is restarted
Add user to group 'libvirt' without relogin
Extract OS distribution tighter
Make dnsmasq configuration only on  Centos
Solve the Libvirt permission problem
Changes made to use of ip commands in ubuntu_bridge_network_configuration.sh and network_teardown_tasks.yml
Update ubuntu_bridge_network_configuration.sh
Added to ip commands

Co-authored-by: Maël Kimmerlin <mael.kimmerlin@est.tech>
Co-authored-by: Kanwar Saad <kanwar.saad.bin.liaqat@est.tech>
Co-authored-by: Jan Tilles <jan.tilles@est.tech>
Co-authored-by: Kashif Khan <kashif.khan@est.tech>

namnx228 committed:
Update checks for CentOS to checks for Redhat-family OS
